### ISSUE TITLE
Improve shared database

### ISSUE THREAD

**Issue by koppor** [koppor/jabref#182]
- [ ] Manage connections: http://discourse.jabref.org/t/managing-shared-database-connections/289
- [ ] Command line access: http://discourse.jabref.org/t/command-line-access-to-sql-databases/288

**Comment by calixtus** [comment]
@s-rousselle: Regarding questions 1-3
https://de.wikipedia.org/wiki/PuTTY

>     4. Should the password of saved databases also be saved or should it be asked upon
>        switching shared databases? If yes, does JabRef have specific guidelines regarding encryption and security of sensitive data?

If the password is stored, jabref shouldn't ask for, except the password is wrong (provide a dialog with save password checkbox)
Until now, jabref stored the proxy password in plain text. Not good, but there wasn't much effort yet to improve that. If you have some time to implement it, go for it.

>     5. Is there any existing remote databases we can use to test the features once implemented or should
>        we set up our own? In the second case, is there anything we should be aware of
>        (e.g. required tables, specific database structure, ...) for JabRef to work properly?

Good questions. Wouldn't be to bad for testing issues, if we had a shared database we can test on. We'll talk about that in the next devcall. Meanwhile, set up your own remote database for testing.

@saschatran: You should be able to move the synchronizing stuff to a background task (jabref already offers infrastructure to follow these thanks to @btut). If there are parts of the task that would block the user from working on the main table, JabRef should display some kind of waiting animation. If there is a huge load of entries (10.000 or so, yes, we really have users with this number of entries). If there is an infinite loop it would probably be a bug, that needs fixing. Maybe @koppor can specify further.

@rohailg24 I believe the task would be to expand the existing cli of jabref to work with shared databases.

**Comment by btut** [comment]
> You should be able to move the synchronizing stuff to a background task

I think the best thing to do here is to go for the same approach as for the pdf indexer. Have one background task with a queue of database queries. You can checkout the [IndexingTaskManager](https://github.com/JabRef/jabref/blob/7eb7b86203d7c149d3b48df8e0845d35aff1597c/src/main/java/org/jabref/logic/pdf/search/indexing/IndexingTaskManager.java) to get an idea of how to do that.

**Comment by rohailg24** [comment]
Hi Thank you for your answer.
According to the instructions, we should automate the process of entries retrieval, is the task to import a Latex Document with all the entries in it or is a simple commandline output sufficient?
Regarding the export, what exactly does the user want to export? a Latex-Document? Or a simple input in the commandline

**Comment by calixtus** [comment]
I don't think that you understood the instructions right...

JabRef works on `.bib`, not with latex documents. It already can sync via the gui entries with a remote database. Locally those entries are stored in a .bib` file, remotely in a database. Syncing/importing/exporting already works in the gui and does not have to be improved (well, actually it could be improved, but thats not the concern of this issue).

This issue applies to improving
* the management of multiple databases (to switch between multiple database, to store credentials, to switch between different profiles etc.)
* making the already existing functionality of the gui accessible via the cli (syncing a `.bib` file with a remote database, running cleanup-formatters, etc.)
* improving the responsiveness of the gui while synchronizing a remote database with your local `.bib` file

**Comment by koppor** [comment]
The responsiveness should not be treated in the work of the course.

Thus, two issues remain:

- Manage connections: http://discourse.jabref.org/t/managing-shared-database-connections/289
- Command line access: http://discourse.jabref.org/t/command-line-access-to-sql-databases/288

> 1. Should JabRef save every remote database you connect to, or only the ones you specify that you want to save?

It should save everyone and offer deletion of existing ones.

> 2. Should the user also be able to save a database without connecting to it?

No, JabRef should do automatic saving

> 3. Should the user be able to edit and/or remove already saved databases?

Yes

> 4. Should the password of saved databases also be saved or should it be asked upon
switching shared databases? 

Yes

> If yes, does JabRef have specific guidelines regarding encryption and security of sensitive data?

No. See https://github.com/JabRef/jabref/issues/8055#issuecomment-954615417 for solution proposals. - Please try to use the proposed solution.

> 5. Is there any existing remote databases we can use to test the features once implemented or should
we set up our own? 

Please setup your own. See the first text of the issue: Use https://www.elephantsql.com/.

> In the second case, is there anything we should be aware of (e.g. required tables, specific database structure, ...) for JabRef to work properly?

Did you try to connect to to https://www.elephantsql.com/? This should work instantly. Please try it and report back.

**Comment by koppor** [comment]
> Regarding "Make UI responsive for shared databases"

This is very hard to work on and needs thorough testing. One should use the YourKit Java Profiler (https://www.yourkit.com/).

>    According to the instructions, the UI becomes unresponsive. Is the goal to make the UI responsive to a reasonable amount of entries?

Yes

>    If so, what would be a reasonable amount of entries?

10.000

>    Or is this issue about figuring out where the problem lies in? 

Yez - we do not know about the root cause. We should observe that interaction with a shared database is slow.

**Comment by saschatran** [comment]
> > 1. Is there any existing remote databases we can use to test the features once implemented or should
> >    we set up our own?
> 
> Please setup your own. See the first text of the issue: Use https://www.elephantsql.com/.
> 
> > In the second case, is there anything we should be aware of (e.g. required tables, specific database structure, ...) for JabRef to work properly?
> 
> Did you try to connect to to https://www.elephantsql.com/? This should work instantly. Please try it and report back.

Yes, we did and it works. (We've already sent the question 1., right before we've figured out what the elephantsql does. Sorry for that.)

Thanks for the answers.

**Comment by s-rousselle** [comment]
It seems we are unable to run JabRef from the .bat file. When typing `JabRef.bat` (or without the `.bat` in Ubuntu) in a terminal, we get the following error:

```
Error occurred during initialization of boot layer
java.lang.module.FindException: Module org.jabref not found
```
It is possible that we missed a step, since we tried that directly in the `build/scripts/` folder in IntelliJ. Could we get some help on how to run JabRef in CLI mode ?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@s-rousselle generally speaking I'd suggest asking these kinds of questions in the Gitter, you are likelier to get a quick response there.

I don't know the goals of your course, but if you want a quick fix that might allow you to proceed until someone gets back to you with a better method, I'd say use `gradlew run` with `--args` instead (e.g., ` ./gradlew run --args="-v"`).

I think @btut did some CLI-related improvements, do you have a better workflow?

**Comment by btut** [comment]
Hi!
You mention you are using intellij. If you also use the Intellij build system you can create a run configuration to test the CLI.
I never used self generated binaries explicitly, but you could try if the pre-built binaries work for you. If they do, there may be a problem with your build.

**Comment by ThiloteE** [comment]
> It seems we are unable to run JabRef from the .bat file. When typing `JabRef.bat` (or without the `.bat` in Ubuntu) in a terminal [...]

How to run JabRef from commandline: https://docs.jabref.org/advanced/commandline

Specifically:
> Do not use JabRef\JabRef.exe or bin/JabRef

Hence, locate the following folder and open the terminal:

`\...\JabRef\lib\runtime\bin\`

Then for Linux:

run `JabRef`

Then for Windows:

run `JabRef.bat`

Assumption: The documentation is up to date.

Not sure if there have to be done special Shenanigans in Intellij
